lib/spack/docs/repositories.rst

```diff
@@ -0,0 +1,275 @@
+repository**, a directory that Spack searches when it needs to find a
+package by name.  By default, the :ref:`cmd-spack-edit` and
+:ref:`cmd-spack-create` commands modify packages in the builtin
+repository.  You may need to maintain a set of packages **separate** from
+the builtin repository, for private, proprietary, export-controlled, or
+otherwise restricted software. Spack allows you to configure local
+repositories using either the ``repos.yaml`` or the ``spack repo``
+command.
+
+A package repository is simply a directory structured like this::
+
+  repo/
+      repo.yaml
+      packages/
+          hdf5/
+              package.py
+          mpich/
+              package.py
+              mpich-1.9-bugfix.patch
+          trilinos/
+              package.py
+          ...
+
+The top-level ``repo.yaml`` file contains configuration metadata for the
+repository, and the ``packages`` directory contains subdirectories for
+each package in the repository.  Each package directory contains a
+``package.py`` file and any patches or other files needed to build the
+package.
+
+
+---------------------
+``repos.yaml``
+---------------------
+
+Spack finds repositories by looking at the ``repos.yaml`` configuration
+file.  For more on the YAML format, and on how configuration file
+precedence works in Spack, :ref:`see configuration <configuration>`.  By
+default, ``repos.yaml`` looks like this:
+
+.. code-block:: yaml
+
+  repos:
+  - $spack/var/spack/repos/builtin
+
+The file starts with ``repos:`` and contains a single ordered list
+(denoted by ``-``).  You can add a repository by inserting another
+element into this list:
+
+.. code-block:: yaml
+
+  repos:
+  - /opt/local-repo
+  - $spack/var/spack/repos/builtin
+
+When you install a package with Spack, it searches these directories in
+order for an appropriate ``package.py`` file.  For example, if you type
+``spack install mpich``, Spack looks first in
+``/opt/local-repo/packages/mpich``, then in
+``$spack/var/spack/repos/builtin/packages/mpich``, and uses the first
+valid package it finds.
+
+.. _note:
+
+  Currently, Spack only knows how to deal with repositories in the local
+  file system. Eventually we plan to support putting URLs in
+  ``repos.yaml`` so that you can easily host remote package repositories,
+  but that support is not implemented yet.
+
+---------------------
+Namespaces
+---------------------
+
+Every repository in Spack has an associated **namespace** defined in its
+top-level ``repo.yaml`` file.  The namespace serves only to distinguish
+packages from different repositories.  If you look at the contents of
+``var/spack/repos/builtin/repo.yaml`` in the builtin repository, you'll
+see that its namespace is ``builtin``:
+
+.. code-block:: console
+
+  $ cat var/spack/repos/builtin/repo.yaml
+  repo:
+    namespace: builtin
+
+The namespace doesn't **have** to correspond to the repository directory
+name; it's just convenient in this case.
+
```

Comment:
As far as upper-case vs. lower-case, I don't think it really matters. It's up to the user. No one else will ever even see it, so no one can complain. No reason to have to enforce this.

> The namespace doesn't **have** to correspond to the repository directory name

Is there any need for this feature? Is the `repo.yaml` only necessary to provide the name of the repo? Why not just use the directory name in all cases?